### PR TITLE
Kill custom threadpools

### PR DESCRIPTION
**Pull request by whirm**

### CODE REVIEW

**Comment by NielsZeilemaker** [comment]
I'm guessing this is not ok
```
GUIDBProducer: DEBUG: GUIDBHandler: Task(tribler_main.py:574 (set_reputation)) took to be called 0.0 (expected 5.0), actual task took 0.0 guiTaskQueue
```
expected 5s, called in 0s.

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 1]
```diff
@@ -831,23 +830,23 @@ def network_get_state(self, usercallback, getpeerlist, sessioncalling=False):
...
-            if sessioncalling:
+            if usercallback:
+                # Invoke the usercallback function via a new thread.
+                # After the callback is invoked, the return values will be passed to the
+                # returncallback for post-callback processing.
+                if not self.done:
+                    # runs on the reactor
+                    def session_getstate_usercallback_target():
+                        (when, getpeerlist) = usercallback(ds)
+                        with self.dllock:
```
Comment (NielsZeilemaker) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: this one doesn't make a lot of sense anymore

File: Tribler/Core/Session.py [hunk 2]
```diff
@@ -364,17 +363,17 @@ def add_observer(self, func, subject, changeTypes=[NTFY_UPDATE, NTFY_INSERT, NTF
...
         """ Remove observer function. No more callbacks will be made.
         @param func The observer function to remove. """
         # Called by any thread
-        self.uch.notifier.remove_observer(func)  # already threadsafe
+        self.notifier.remove_observer(func)  # already threadsafe
 
     def open_dbhandler(self, subject):
         """ Opens a connection to the specified database. Only the thread
-        calling this method may use this connection. The connection must be
+        calling this method may use thisf connection. The connection must be
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: typo

File: Tribler/Main/Utility/GuiDBHandler.py [hunk 3]
```diff
@@ -79,7 +75,7 @@ def onSameThread(self, type):
         if type == "dbThread" or isInIOThread():
             return isInIOThread()
 
-        return threading.currentThread().getName().startswith('GUITaskQueue')
+        return False
```
Comment (NielsZeilemaker) [inline Tribler/Main/Utility/GuiDBHandler.py]: the IOThread is the twisted thread, but this return statement should check if we're in a twisted thread pool

File: Tribler/Main/Utility/GuiDBHandler.py [hunk 4]
```diff
@@ -166,40 +162,33 @@ def wrapper():
         wrapper.__name__ = str(name)
 
         if not self.onSameThread(workerType) or delay:
+            task_name = uId
```
Comment (NielsZeilemaker) [inline Tribler/Main/Utility/GuiDBHandler.py]: The add_task will generate a task_name if name is None, I guess this can be removed.

**Comment by NielsZeilemaker** [comment]
Its looking really nice @whirm

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py (hunk 1 above)
Comment (whirm) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: You mean acquiring the lock?

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by NielsZeilemaker** [comment]
The failure was caused by a pydeadobject

**Comment by NielsZeilemaker** [comment]
@whirm, it's not a problem if we want to run in the threadpool but are running in the twisted thread. 
So I guess guidbhandler needs to change slightly.

**Comment by whirm** [comment]
> @whirm, it's not a problem if we want to run in the threadpool but are running in the twisted thread.
> So I guess guidbhandler needs to change slightly.

As in calling GuiDBHandler.Add() from the reactor thread to schedule something on the threadpool?

File: Tribler/Main/Utility/GuiDBHandler.py [hunk 5]
```diff
@@ -165,41 +156,34 @@ def wrapper():
...
 
-        if not self.onSameThread(workerType) or delay:
-            if workerType == "dbThread":
-                task_name = uId
-                if not task_name:
-                    self._auto_counter += 1
-                    task_name = "guidbhandler %d" % self._auto_counter
+        db_thread_wanted = workerType == "dbThread"
+        threadpool_wanted = workerType == "guiTaskQueue"
+        can_run_in_same_thread = (db_thread_wanted and isInIOThread()) or (threadpool_wanted and isInThreadPool())
```
Comment (NielsZeilemaker) [inline Tribler/Main/Utility/GuiDBHandler.py]: Here, the onSameThread method would check if we're on the IOThread if workerType == "dbThread".
But if we're on the IOThread, it doesn't care about the workerType. This reduces latency slightly.

Comment (whirm) [inline Tribler/Main/Utility/GuiDBHandler.py]: Oh! so onSameThread meant "we are going to call the stuff on the same thread" as opposed to "we are on the same thread we want to schedule the task in"?

So if we are on the reactor thread, we disregard the request to be executed on the threadpool as the only thing it wants really is to run out of the GUI thread?

Comment (NielsZeilemaker) [inline Tribler/Main/Utility/GuiDBHandler.py]: Jup

Comment (whirm) [inline Tribler/Main/Utility/GuiDBHandler.py]: Oh man, that's nuts and undocumented :D. I'll have a shot at clarifying the code a bit further then.

Comment (NielsZeilemaker) [inline Tribler/Main/Utility/GuiDBHandler.py]: It's only used in the gui, so it should be fine...

**Comment by whirm** [comment]
Ok, now we just check if we are on the threadpool or the reactor thread.

If the rest looks fine to you, I can squash and then it can be merged.

**Comment by NielsZeilemaker** [comment]
Except for this last bug, everything looks fine.

**Comment by whirm** [comment]
Disregard that last message the chat refreshed when I sent it and it wasn't showing your last one.

> I'm not sure if this is ok, because if we're in the threadpool then we might need to switch to the iothread.

```add_task()`` switches to the reactor thread to schedule the delayed task which will be executed on the reactor thread.

**Comment by NielsZeilemaker** [comment]
But if there's no delay, we might execute a task on the threadpool destined for the iothread.

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
@NielsZeilemaker @LipuFei sould be ready.

**Comment by LipuFei** [comment]
@whirm @NielsZeilemaker Looks good.

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
```One test raised 1 unhandled exceptions, last one was: AssertionError('no channels matching tv',)```
http://jenkins.tribler.org//job/GH_Tribler_pull-request-tester_devel/4436/testReport/junit/Tribler.Test.test_remote_search/TestRemoteChannelSearch/test_channel_search/

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by synctext** [comment]
Getting quite stable, awsome!. recent fail on "Time out, killing monitored processes...."

```
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: 	 COVERAGE_FILE=.coverage.9 wrap_in_vnc.sh 'nosetests -v --with-coverage  --xunit-file=/home/jenkins/workspace/GH_Tribler_pull-request-tester-turbo_devel/output/9_nosetests.xml --with-xunit --all-modules --traverse-namespace --cover-package=Tribler --cover-inclusive  'Tribler/Test/API/test_seeding.py Tribler/Test/API/test_seeding_vod.py Tribler/Test/test_hidden_community.py
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: 	 exited with status: 0
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: Time out, killing monitored processes.
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: TERMinating group. Still 1 process(es) running:
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT:  - /bin/sh-cCOVERAGE_FILE=.coverage.5 wrap_in_vnc.sh 'nosetests -v --with-coverage  --xunit-file=/home/jenkins/workspace/GH_Tribler_pull-request-tester-turbo_devel/output/5_nosetests.xml --with-xunit --all-modules --traverse-namespace --cover-package=Tribler --cover-inclusive  'Tribler/Test/test_tunnel_community_fails.py Tribler/Test/test_remote_search.py Tribler/Test/test_i2i.py
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: Command:
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: 	 COVERAGE_FILE=.coverage.5 wrap_in_vnc.sh 'nosetests -v --with-coverage  --xunit-file=/home/jenkins/workspace/GH_Tribler_pull-request-tester-turbo_devel/output/5_nosetests.xml --with-xunit --all-modules --traverse-namespace --cover-package=Tribler --cover-inclusive  'Tribler/Test/test_tunnel_community_fails.py Tribler/Test/test_remote_search.py Tribler/Test/test_i2i.py
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: 	 exited with status: -15
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] OUT: ERROR: Process guard failed with exit code 3, aborting and printing logs
INFO:OneShotProcessProtocol:[run_nosetests_for_je...] ERR: /home/jenkins/workspace/GH_Tribler_pull-request-tester-turbo_devel/gumby/scripts/run_nosetests_for_jenkins.sh: line 111: kill: (15472) - No such process
INFO:OneShotProcessProtocol:[run_nosetests_for_jenkins.sh] exit code 1
ERROR:ExperimentRunner:Running the experiment instances failed, collecting data before failing.
INFO:ExperimentRunner:Syncing output data back from head nodes...
INFO:ExperimentRunner:Great copying success!
ERROR:ExperimentRunner:Experiment execution failed, exiting with error.
ERROR:ExperimentRunner:<twisted.python.failure.Failure <class 'twisted.internet.defer.FirstError'>>
Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 416, in fireEvent
    DeferredList(beforeResults).addCallback(self._continueFiring)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 306, in addCallback
    callbackKeywords=kw)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 295, in addCallbacks
    self._runCallbacks()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 577, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 429, in _continueFiring
    callable(*args, **kwargs)
exceptions.SystemExit: 1
```

**Comment by whirm** [comment]
@synctext It's not related to this PR, one of the remote search tests gets stuck sometimes, when this one is merged, I'll rebase and continue my work on the branch dedicated to that.

**Comment by whirm** [comment]
retest this please

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 6]
```diff
@@ -833,23 +832,22 @@ def network_get_state(self, usercallback, getpeerlist, sessioncalling=False):
...
+                # returncallback for post-callback processing.
+                if not self.done:
+                    # runs on the reactor
+                    def session_getstate_usercallback_target():
+                        when, getpeerlist = usercallback(ds)
+                        if when > 0.0:
+                            # Schedule next invocation, either on general or DL specific
+                            self.session.lm.rawserver.add_task(lambda: self.network_get_state(usercallback, getpeerlist), when)
+
+                    callInThreadPool(session_getstate_usercallback_target)
```
Comment (NielsZeilemaker) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: Why did you choose to go around the new rawserver, and call callInThreadPool directly?

Comment (whirm) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: Hm, that's a mixup. I'll fix it. Thanks.

Comment (NielsZeilemaker) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: It's happening in other places as well, I think like 6 times in total.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 7]
```diff
@@ -340,7 +340,7 @@ def get_metainfo(self, infohash_or_magnet, callback, timeout=30, timeout_callbac
 
             cache_result = self._get_cached_metainfo(infohash)
             if cache_result:
-                self.trsession.uch.perform_usercallback(lambda cb=callback, mi=deepcopy(cache_result): cb(mi))
+                callInThreadPool(callback, deepcopy(cache_result))
```
Comment (NielsZeilemaker) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: Here are a couple more

File: Tribler/Core/RemoteTorrentHandler.py [hunk 8]
```diff
@@ -226,7 +227,7 @@ def notify_possible_torrent_infohash(self, infohash):
             return
 
         for callback in self.torrent_callbacks[infohash]:
-            self.session.uch.perform_usercallback(lambda ucb=callback, ih=hexlify(infohash): ucb(ih))
+            callInThreadPool(callback, hexlify(infohash))
```
Comment (NielsZeilemaker) [inline Tribler/Core/RemoteTorrentHandler.py]: And here

**Comment by whirm** [comment]
@NielsZeilemaker yes, I'm testing the changes locally.

Re bypassing the TwistedRawserver, I now remember I thought that as they where just scheduling stuf for immediate call on the thread pool and, as the TaskManager can't cancel tasks that are already active It didn't make a lot of sense to go that route anyway. 
But now I'm thinking that I'll extend the RawServer so we can wait for tasks that are still in execution on the thread pool at shutdown time. I'll do that in a different PR but I'll go back to using the RawServer in this one. I'm also adding a few utility methods to the RawServer so we can schedule stuff without having to create all those lambdas.

**Comment by NielsZeilemaker** [comment]
Great idea

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
@NielsZeilemaker how doest it look to you now? Should I squash fixups so it can be merged?

**Comment by whirm** [comment]
The job's timeout threshold got set too low due to @rjruigrok improvements on the tunnel tests.

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
@LipuFei @NielsZeilemaker ready

**Pull request merged**